### PR TITLE
Properly align the stack pointer

### PR DESCRIPTION
**Pull request by dschuff**
By default (if no global base is given) the global base is 1, which
seems wrong. In this case the stack pointer gets an address of 1, which
is unaligned and definitely wrong. So, start the global base at 0 instead of
1 by default and align the stack pointer. Also factor allocation of
statics into a function.

### CODE REVIEW

**Comment by dschuff** [comment]
Forked from #316

**Comment by dschuff** [comment]
The current (and previous) behavior is that the stack pointer would always be allocated even if the stack was not allocated. This was because the compiler will always emit references to it if necessary (e.g. in tests like `test/llvm_autogenerated/byval.s`) without knowing if a stack is going to be allocated or not; and if there was not at least a stack pointer allocated, the test would fail to link. The fact that the tests previously had no memory was a bug. Practically speaking this only affects the binaryen tests because everywhere we actually use s2wasm that I know of, we use the `--allocate-stack` flag.
So, we have a couple of options.

1. Make the stack pointer only allocated if a stack is being allocated. If we do this we will need to use the `--allocate-stack` flag for some or all of the binaryen tests. If we really want to have tests with no memory, we'll have to update check.py to use `--allocate-stack` for only some of the tests.
2. Make the pointer allocated unconditionally and just accept that all of our tests will have memory.

I don't particularly care which of those we do; do you have an opinion?

File: src/s2wasm.h [hunk 1]
```diff
@@ -420,32 +420,31 @@ class S2WasmBuilder {
     }
   }
 
+  // Allocate a static variable and return its address in linear memory
+  size_t allocateStatic(size_t allocSize, size_t alignment, Name name) {
+    size_t address = (nextStatic + alignment - 1) & static_cast<size_t>(-static_cast<ssize_t>(alignment));
```
Comment (jfbastien) [inline src/s2wasm.h]: Assert alignment is a power of 2?

Comment (jfbastien) [inline src/s2wasm.h]: Add to support/ utilities similar to LLVM's:
```
inline uintptr_t alignAddr(const void *Addr, size_t Alignment) {
  assert(Alignment && isPowerOf2_64((uint64_t)Alignment) &&
         "Alignment is not a power of two!");

  assert((uintptr_t)Addr + Alignment - 1 >= (uintptr_t)Addr);

  return (((uintptr_t)Addr + Alignment - 1) & ~(uintptr_t)(Alignment - 1));
}

/// \brief Returns the necessary adjustment for aligning \c Ptr to \c Alignment
/// bytes, rounding up.
inline size_t alignmentAdjustment(const void *Ptr, size_t Alignment) {
  return alignAddr(Ptr, Alignment) - (uintptr_t)Ptr;
}
```
Nicely reusable.

**Comment by kripken** [comment]
1. No strong opinions. But I'm still not sure why fixing the *alignment* of the stack pointer causes memory to be used, when it wasn't before. Weren't we allocating it before, just unaligned?

2. I still think that 1 should be 0. Otherwise the first global allocated might be at address 0, which is an aligned address...

File: src/s2wasm.h (hunk 1 above)
Comment (dschuff) [inline src/s2wasm.h]: Done.

**Comment by dschuff** [comment]
Fixing the alignment didn't cause memory to be used, per se.
I guess another way of describing the previous behavior was that we reserved an address for the stack pointer by bumping `nextStatic` but we didn't also finish the allocation by also bumping `minInitialMemory` if there was no stack allocation. So if there were no other statics or anything else that required reserving memory, then no memory would be allocated.

Actually come to think of it, now that the allocation happens here in the function, I don't think we actually need separate variables for `nextStatic` and `minInitialMemory`. I'll fix that.

So we still have the 2 options of conditionally or unconditionally allocating the stack pointer. I'm leaning toward option 1 (so the program will fail to link if there is a reference to the stack pointer but no stack allocation) and just using `--allocate-stack` for the llvm-autogenerated tests.

Do you mean the 0 should be 1, i.e. the first static should be allocated at address 1 (assuming no alignment constraints)? That would mean that nothing could ever be allocated at 0. And practically speaking if there is a stack pointer it would push its address up to 4, wasting the first 4 bytes.

**Comment by dschuff** [comment]
OK, I made it so the stack pointer is only allocated when the stack is allocated. For testing we have the hand-created `dot_s` tests and the `llvm_autogenerated` tests. For the former I kept it with no stack allocation (none of them use the stack pointer), and for the latter I made the test runner pass the `--allocate-stack` flag. If you look at the test updates, you'll see that none of the `dot_s` tests that didn't have any memory before have any memory now, and that some of the `llvm_autogenerated` tests such as `byval.s` (which previously had no memory even though they referenced the stack pointer) are now more correct. Also you'll see that making the static base 0 instead of 1 reduced the memory requirements of a lot of tests (obviously they still need a whole page, but the addresses used are smaller).

**Comment by dschuff** [comment]
Or wait, are you saying that we should just intentionally disallow anything from being allocated at address 0?

**Comment by kripken** [comment]
Yeah, I think 0 should never be an ok address to allocate on, so we start at 1, which gets aligned for the first one to something > 0.

**Comment by dschuff** [comment]
But what's the point of disallowing 0 if you can allocate at 4? That's not useful for any purpose I can think of (e.g. catching null pointer derefs).

**Comment by dschuff** [comment]
oh, also. It looks like emscripten is not using the `--allocate-stack` flag, which means the stack pointer appears to be uninitialized in the wast file. I assume it wants to tie it to STACKTOP, where does that happen?

**Comment by kripken** [comment]
I don't follow, what's wrong with 4? 0 is bad because it's the same as `nullptr`, so it would break comparisons to null.

Emscripten will set up the stack itself. It emits a little code that writes the var `STACKTOP` into the memory location for the stack pointer, over [here](https://github.com/kripken/emscripten/blob/incoming/emscripten.py#L1532).

**Comment by dschuff** [comment]
Oh, I found it, emscripten assumes that it lives at the `--global-base` location, which will be true if that location is aligned. So that means that conditionally allocating the stack pointer won't work for emscripten. So I guess we can just go back to option 2 above.

**Comment by kripken** [comment]
Sounds good.

**Comment by dschuff** [comment]
Ok. s2wasm now explicitly reserves address 0 (instead of making it depend on getting 1 passed to the constructor for `globalBase`) and unconditionally allocates the stack pointer. This means that s2wasm will never generate wasts with no memory, but since it's designed to work with LLVM output and LLVM may use the stack pointer anytime, and because emscripten always does its own stack, this seems ok to me.

File: src/s2wasm.h [hunk 2]
```diff
@@ -58,7 +58,6 @@ class S2WasmBuilder {
         startFunction(startFunction),
         globalBase(globalBase),
         nextStatic(globalBase),
-        minInitialMemory(0),
```
Comment (kripken) [inline src/s2wasm.h]: are you sure this gets written to before it's read? I think that's not the case, and we do need this here (for when there is no stack, no globals, etc.).

Comment (dschuff) [inline src/s2wasm.h]: `minInitialMemory` is now completely gone. In the previous commit it was always updated exactly when `nextStatic` was updated and exactly to its value. It only had one other use (at the top of `fix()` which now uses `nextStatic` instead. The consequence of unconditionally allocating the stack pointer is that there will never be a case with no globals.

Comment (kripken) [inline src/s2wasm.h]: Oh right, sorry. I was reading it wrong.

**Pull request merged**